testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java

```diff
@@ -0,0 +1,318 @@
+                    int offset = -1;
+                    for (int i = 0; i < processed.size(); i++) {
+                        // if there is a room, advance the offset
+                        if (offset < produced.size() - 1) {
+                            offset++;
+                        }
+
+                        T x = produced.get(offset);
+                        T y = processed.get(i);
+                        if (!x.equals(y)) {
+                            // if processing task doesn't match to produced task, it means re-consuming happens.
+                            // lookup the offset to rewind
+                            Integer lookup = toIndex.get(y);
+
+                            // rewound offset cannot be greater than current offset
+                            if (lookup == null || lookup > offset) {
+                                return false;
+                            }
+                            offset = lookup;
+                        }
+                    }
+                    // must be processed until the end of produced tasks
+                    return offset == produced.size() - 1;
+            }
+            throw new RuntimeException("Unreachable");
+        }
+    }
+
+    @Value
+    @Accessors(fluent = true)
+    public static class DecatonProducerRecord<T> {
+        String key;
+        T task;
+    }
+
+    @Setter
+    @Accessors(fluent = true)
+    public static class Builder<T extends MessageLite> {
+        private final KafkaClusterRule rule;
+        private final Parser<T> parser;
+
+        /**
+         * Number of subscription instances which consumes tasks
+         */
+        private int numSubscriptionInstances = 3;
+        /**
+         * Configure test-specific processing logic
+         */
+        private Function<ProcessorsBuilder<T>, ProcessorsBuilder<T>> configureProcessorsBuilder;
+        private RetryConfig retryConfig;
+        private PropertySupplier propertySupplier;
+        /**
+         * Expected process ordering for this test
+         */
+        private ProcessOrdering expectOrdering = ProcessOrdering.ALLOW_DUPLICATE;
+
+        @Setter(AccessLevel.NONE)
+        private int numTasks;
+        @Setter(AccessLevel.NONE)
+        private Iterator<DecatonProducerRecord<T>> taskIterator;
+
+        private Builder(KafkaClusterRule rule, Parser<T> parser) {
+            this.rule = rule;
+            this.parser = parser;
+        }
+
+        /**
+         * Configure tasks which are produced to the topic.
+         * Note that all tasks should be differentiated from others (i.e. equals() never be true between tasks)
+         * since message ordering will be verified by checking produced task equals to processed task one by one.
+         *
+         * @param numTasks number of tasks
+         * @param taskIterator iterator which generates tasks
+         */
+        public Builder<T> produce(int numTasks,
```

Comment:
I've also considered what's the most suitable interface for this "integration test framework" and summarized stuff like this:

Input(parameters) needs to be customizable:
* How `ProcessorSubscription` is constructed
* What `DecatonProcessor` does (e.g, retry(), throw exception etc)
* List of semantics expected to be achieved  with given setup (e.g, ordering guarantee is required by default but is no longer valid when retry is enabled)

Checks (those should be assumed by default, but should be possible to opt-out as the setup confirms to relax):
* Process ordering based on keys
* Duplicating tasks
* Missing tasks
* Serial processing (no two tasks sharing the same key gets processed at the same time)
* and maybe more to add

Other than those things, including the number of tasks to generate or what to include in each tasks are often not a matter. Given that, I think this framework can generate tasks with arbitrary default numbers by itself. (It would be possibly good to leave a room to add `opaque` in the task though. e.g, `.taskData(T opaque)`.

So first of all, I don't see necessity of letting users of this framework to fully custom generated tasks.

In addition, I think it's better that each checks are encupslated into a interface like `interface Check { ... } ` which holds methods to get checked-in on each event (e.g, receive task) and make them possible to opt-out with interface like `Suite.builder().optOutChecks(Checks.DELIVERY_ORDER)` or so.

Reply:
Fair enough.
> I don't see necessity of letting users of this framework to fully custom generated tasks

Sounds nice, but how `interface Check` should be?
> I think it's better that each checks are encupslated into a interface like interface Check { ... }

For example, I think "Serial processing" can be checked by observing each received task (by having per-key mutex e.g.), but to check "Missing tasks", we have to pass "produced tasks" to detect which task is missing.

Reply:
I think it can be something like below, taking a form of observer for each events.
Then it can accumulate contexts internally and use it to check on `assertSemantic`. 
There might be duplicate in information stored in each check but it would be reasonable since each check may requires different data structure to check against them eventually.

```java
    interface Check {
        void onProduce(String key, Task task);

        void onProcess(String key, Task task);

        void assertSemantic();
    }
```


Regarding "Serial processing" I think it's better to record process being/end time of each records and confirm there're no overlap between two records sharing the same key.

Reply:
Much simpler, looks good
> confirm there're no overlap between two records